contracts/root/staking/CustomSupernetManager.sol

```diff
@@ -129,9 +134,31 @@ contract CustomSupernetManager is ICustomSupernetManager, Ownable2StepUpgradeabl
         return validators[validator_];
     }
 
+    /**
+     *
+     * @inheritdoc ICustomSupernetManager
+     */
+    function premine(uint256 amount) external {
+        if (address(_rootERC20Predicate) == address(0)) {
+            revert Unauthorized("CustomSupernetManager: UNDEFINED_ROOT_ERC20_PREDICATE");
+        }
+
+        IERC20 nativeTokenRoot = IERC20(_rootERC20Predicate.nativeTokenRoot());
+        if (address(nativeTokenRoot) == address(0)) {
+            revert Unauthorized("CustomSupernetManager: UNDEFINED_NATIVE_TOKEN_ROOT");
+        }
+        require(!_genesis.completed(), "CustomSupernetManager: GENESIS_SET_IS_ALREADY_FINALIZED");
+
+        nativeTokenRoot.safeTransferFrom(msg.sender, address(_rootERC20Predicate), amount);
```

Comment:
Who is supposed to call this functions? Asking because `msg.sender` is referenced, but the requirement before was that genesis was not completed, so I'm wondering if the `msg.sender` can actually call it. A bit of info here appreciated.

Reply:
The approach is the same as with staking of genesis validators. Each validator, before genesis set is finalized, invokes `StakeManager.stakeFor`, which then invokes `CustomSupernetManager._onStake` (https://github.com/Stefan-Ethernal/core-contracts/blob/579029fc472eec211996b4313c1ed675962417bd/contracts/root/staking/CustomSupernetManager.sol#L161).

So basically we are able to do the premine for any account, but only before genesis set gets finalized (during bootstrap basically) and never again in the runtime. In contrary to staking which is able to be invoked either during Supernets bootstrapping (for the genesis validators), but also once Supernet is up and running. 

Also to mention that it is allowed to invoke premine function only in case when rootchain token is used as the Supernets native token.